packages/image_picker/image_picker_for_web/lib/image_picker_for_web.dart

```diff
@@ -0,0 +1,182 @@
+    String capture;
+    if (source == ImageSource.camera) {
+      capture = device == CameraDevice.front ? 'user' : 'environment';
+    }
+    return capture;
+  }
+
+  html.File _getFileFromInput(html.FileUploadInputElement input) {
+    if (_hasOverrides) {
+      return _overrides.getFileFromInput(input);
+    }
+    return input.files[0];
+  }
+
+  /// Handles the OnChange event from a FileUploadInputElement object
+  /// Returns the objectURL of the selected file.
+  String _handleOnChangeEvent(html.Event event) {
+    // load the file...
+    final html.FileUploadInputElement input = event.target;
+    final html.File file = _getFileFromInput(input);
+
+    if (file != null) {
+      return html.Url.createObjectUrl(file);
+    }
+    return null;
+  }
+
+  /// Monitors an <input type="file"> and returns the selected file.
+  Future<PickedFile> _getSelectedFile(html.FileUploadInputElement input) async {
+    // Observe the input until we can return something
+    final Completer<PickedFile> _completer = Completer<PickedFile>();
+    input.onChange.listen((html.Event event) async {
+      final objectUrl = _handleOnChangeEvent(event);
+      _completer.complete(PickedFile(objectUrl));
+    });
+    input.onError // What other events signal failure?
+        .listen((html.Event event) {
+      _completer.completeError(event);
+    });
+
+    return _completer.future;
+  }
+
+  /// Initializes a DOM container where we can host input elements.
+  html.Element _initTarget(String id) {
+    var target = html.querySelector('#${id}');
+    if (target == null) {
+      final html.Element targetElement =
+          html.Element.tag('flt-image-picker-inputs')..id = id;
+
+      html.querySelector('body').children.add(targetElement);
+      target = targetElement;
+    }
+    return target;
+  }
+
+  /// Creates an input element that accepts certain file types, and
+  /// allows to `capture` from the device's cameras (where supported)
+  @visibleForTesting
+  html.Element createInputElement(String accept, String capture) {
+    if (_hasOverrides) {
+      return _overrides.createInputElement(accept, capture);
+    }
+
+    html.Element element;
+
+    if (capture != null) {
+      // Capture is not supported by dart:html :/
+      element = html.Element.html(
+          '<input type="file" accept="$accept" capture="$capture" />',
+          validator: html.NodeValidatorBuilder()
+            ..allowElement('input', attributes: ['type', 'accept', 'capture']));
+    } else {
+      element = html.FileUploadInputElement()..accept = accept;
+    }
+
+    return element;
+  }
+
+  /// Injects the file input element, and clicks on it
+  void _injectAndActivate(html.Element element) {
+    _target.children.clear();
+    _target.children.add(element);
+    element.click();
+  }
+}
+
+// Some tools to override behavior for unit-testing
+typedef _OverrideCreateInputFunction = html.Element Function(
```

Comment:
These should be public as well if they are visible to tests. It's generally an antipattern for a public interface to vent private types. `@visibleForTesting` is a fair game though.

Reply:
Since neither the end user nor the tests need this as a left-hand assignment on their own, I left them private, but I can make them public and @visibleForTesting. At least that way they'll have documentation :)

Reply:
Yeah, I know that the analyzer leaks the private type to the call site in this case, but it still feels weird to vend private types publicly.